pydantic/_internal/_decorators.py

```diff
@@ -218,3 +229,106 @@ def in_ipython() -> bool:
+        ...
+
+
+class V1ValidatorWithValuesKwOnly(Protocol):
+    def __call__(self, __value: Any, *, values: dict[str, Any]) -> Any:  # pragma: no cover
+        ...
+
+
+class V1ValidatorWithKwargs(Protocol):
+    def __call__(self, __value: Any, **kwargs: Any) -> Any:  # pragma: no cover
+        ...
+
+
+class V1ValidatorWithKwargsAndValue(Protocol):
+    def __call__(self, __value: Any, values: dict[str, Any], **kwargs: Any) -> Any:  # pragma: no cover
+        ...
+
+
+class V1ValidatorWithValuesAndKwargs(Protocol):
+    def __call__(self, __value: Any, values: dict[str, Any], **kwargs: Any) -> Any:  # pragma: no cover
+        ...
+
+
+V1Validator = Union[
+    V1ValidatorWithValues, V1ValidatorWithValuesKwOnly, V1ValidatorWithKwargs, V1ValidatorWithValuesAndKwargs
+]
+
+
+def make_generic_validator(
+    validator: V1Validator | OnlyValueValidator | ValidatorFunction, mode: str
+) -> ValidatorFunction:
+    """
+    Make a generic function which calls a validator with the right arguments.
+    """
+    sig = signature(validator)
+
+    positional_params: List[str] = []
+    keyword_only_params: List[str] = []
+    accepts_kwargs = False
+    for param_name, parameter in sig.parameters.items():
+        if param_name in ('field', 'config'):
+            raise TypeError(
+                'The `field` and `config` parameters are not available in Pydantic V2.'
+                ' Please use the `info` parameter instead.'
+                ' You can access the configuration via `info.config`,'
+                ' but it is a dictionary instead of an object like it was in Pydantic V1.'
+                ' The `field` argument is unavailable.'
+            )
+        if parameter.kind in (Parameter.POSITIONAL_ONLY, Parameter.POSITIONAL_OR_KEYWORD):
+            positional_params.append(param_name)
+        elif parameter.kind is Parameter.KEYWORD_ONLY:
+            keyword_only_params.append(param_name)
+        elif parameter.kind is Parameter.VAR_KEYWORD:
+            accepts_kwargs = True
+
+    accepts_values_kw = (keyword_only_params == ['values'] and len(positional_params) == 1) or (
+        len(positional_params) == 2 and positional_params[1] == 'values'
+    )
+
+    if accepts_kwargs and len(positional_params) == 1:
+        # although this could be compatible with the V2 signature we want to discourage it,
+        # so we treat it as a backwards compatible validator
+        validator = cast(V1ValidatorWithKwargs, validator)
+
+        @wraps(validator)
+        def _wrapper1(value: Any, info: ValidationInfo) -> Any:
+            return validator(value, values=info.data)  # type: ignore[call-arg]
+
+        return _wrapper1
+    if len(positional_params) == 1 and keyword_only_params == []:
+        validator = cast(OnlyValueValidator, validator)
+
+        @wraps(validator)
+        def _wrapper2(value: Any, info: ValidationInfo) -> Any:
+            return validator(value)  # type: ignore[call-arg]
+
+        return _wrapper2
```

Comment:
I think this is a candidate for passing an argument along the lines of `accepts_info_param` to Rust just to cut out one more Python function call. Either way, I'd keep this signature option

Reply:
Yup, I'm up for `accepts_info_param` being added to the core-schema.

Reply:
tabling this for a future PR, it's really just a performance opt